openssl/src/x509/mod.rs

```diff
@@ -791,6 +905,75 @@ impl PartialEq<X509Ref> for X509 {
 
 impl Eq for X509 {}
 
+foreign_type_and_impl_send_sync! {
+    type CType = ffi::X509_ATTRIBUTE;
+    fn drop = ffi::X509_ATTRIBUTE_free;
+
+    /// Permit additional fields to be added to an `X509` v3 certificate.
+    pub struct X509Attribute;
+    /// Reference to `X509Attribute`.
+    pub struct X509AttributeRef;
+}
+
+impl Stackable for X509Attribute {
+    type StackType = ffi::stack_st_X509_ATTRIBUTE;
+}
+
+impl X509Attribute {
+    /// Creates an X509 attribute, which can be added to X509 certificates, signing requests and
+    /// PKCS7 messages.
+    ///
+    /// The attribute type is derived from the ASN1 string value. So be sure to use
+    /// Asn1String::type_new() for strings other than octet strings.
+    ///
+    pub fn from_string(nid: Nid, value: Asn1String) -> Result<X509Attribute, ErrorStack> {
+        unsafe {
+            let asn1_type = cvt(ffi::ASN1_STRING_type(value.as_ptr()))?;
+            let attribute = cvt_p(ffi::X509_ATTRIBUTE_create(
+                nid.as_raw(),
+                asn1_type,
+                value.as_ptr() as *mut c_void,
+            ));
+            #[allow(clippy::forget_copy)]
+            mem::forget(value); // Asn1String does not implement the Copy trait, so clippy is wrong here.
+            attribute.map(X509Attribute)
+        }
+    }
+
+    /// Creates an X509 attribute from an `Asn1Object`.
+    pub fn from_object(nid: Nid, value: Asn1Object) -> Result<X509Attribute, ErrorStack> {
+        unsafe {
+            let asn1_type = Asn1TagValue::OBJECT;
+            let attribute = cvt_p(ffi::X509_ATTRIBUTE_create(
+                nid.as_raw(),
+                asn1_type.as_raw(),
+                value.as_ptr() as *mut c_void,
+            ));
+            #[allow(clippy::forget_copy)]
+            mem::forget(value); // Asn1String does not implement the Copy trait, so clippy is wrong here.
+            attribute.map(X509Attribute)
+        }
+    }
+}
+
+impl X509AttributeRef {
+    /// Get object id (NID) of attribute
+    pub fn object(&self) -> Result<&Asn1ObjectRef, ErrorStack> {
+        unsafe {
+            let object_ptr = cvt_p(ffi::X509_ATTRIBUTE_get0_object(self.as_ptr()))?;
+            Ok(Asn1ObjectRef::from_ptr(object_ptr))
+        }
+    }
+
+    /// Get an item from the list of `ASN1_TYPE`s.
+    pub fn typ(&self, idx: isize) -> Result<&Asn1TypeRef, ErrorStack> {
+        unsafe {
+            let type_ptr = cvt_p(ffi::X509_ATTRIBUTE_get0_type(self.as_ptr(), idx as c_int))?;
```

Comment:
This must test the length of the stack first so it doesn't do an out-of-bounds read.